File: 프로그래머스/1/42889. 실패율/실패율.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    vector<vector<float>> user(N+1,vector<float>(2,0));

    for (int num: stages) {
        if (num<=N){
            user[num][0]+=1;    
        }
        if (num>N){
            num=N;
        }
        for (int j=1;j<=num;++j){
            user[j][1]+=1;
        }
    }
    for (int i=1;i<=N;++i){
        if (user[i][1] == 0) user[i][0] = 0;  
        else user[i][0] /= user[i][1];
    }
    for (int i=1;i<=N;++i){
        answer.push_back(i);
    }

    for (int i=0; i<N-1; ++i) {
        int idx = i;
        for (int j=i+1; j<N; ++j) {
            if (user[answer[j]][0] > user[answer[idx]][0]) {
                idx = j;
            }
            else if (user[answer[j]][0] == user[answer[idx]][0] && answer[j] < answer[idx]) {
                idx = j;
            }
        }
        if (idx != i) swap(answer[i], answer[idx]);
    }

    return answer;
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    user=[0]*(N+1)
    
    for stage in stages:
        if stage > N: stage=N
        for i in range(1,stage+1):
            user[i]+=1

    for i in range(1,N+1):
        if not user[i]: continue
        user[i]=stages.count(i)/user[i]

    answer=[[v,i+1] for i,v in enumerate(user[1:])]
    
    for idx in range(N-1):
        for j in range(idx+1,N):
            if answer[idx][0] < answer[j][0]: answer[idx],answer[j]=answer[j],answer[idx]
            elif answer[idx][0]==answer[j][0] and answer[idx][1]>answer[j][1]: answer[idx],answer[j]=answer[j],answer[idx]

    return [value[1] for value in answer]